### PR TITLE
Added SimpleExpressionMatcher

### PR DESCRIPTION
**Pull request by sbayer55**
Signed-off-by: Steven Bayer <smbayer@amazon.com>

### Description
SimpleExpressionMatcher as an abstract base matcher for creating grammar matcher classes
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by dapowers87 — DISMISSED** [review]

File: data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/util/ContextMatcher.java [hunk 1]
```diff
@@ -43,6 +43,20 @@
  * </p>
  */
 public class ContextMatcher extends DiagnosingMatcher<ParseTree> {
+    public static void describeContextTo(final ParseTree ctx, final Description mismatch) {
```
Comment (sshivanii) [inline data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/util/ContextMatcher.java:46]: Nit: Can this variable be renamed to `parseTreeContext`?

File: data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/util/SimpleExpressionMatcher.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+import java.util.List;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.opensearch.dataprepper.expression.util.ContextMatcher.describeContextTo;
+
+/**
+ * Base class for Matcher that asserts a unary tree ending in a node that matches {@link SimpleExpressionMatcher#lastNodeClass}
+ */
+abstract class SimpleExpressionMatcher extends DiagnosingMatcher<ParseTree> {
+    private static final Matcher<Integer> childCountMatcher = is(1);
```
Comment (dinujoh) [inline data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/util/SimpleExpressionMatcher.java:22]: consider renaming this to singleChildCountMatcher

File: data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/util/SimpleExpressionMatcher.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+    protected final List<Class<? extends ParseTree>> validRuleOrder;
+    protected final Class<? extends ParseTree> lastNodeClass;
+
+    protected SimpleExpressionMatcher(final List<Class<? extends ParseTree>> validRuleOrder) {
+        this.validRuleOrder = validRuleOrder;
```
Comment (dinujoh) [inline data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/util/SimpleExpressionMatcher.java:28]: Can we also store List of validRuleTypeOrder which store the type ? The getRuleIndex method will be reduced to below (you also wouldn't require this private method):

```
private int getRuleIndex(final ParseTree item) {
 return validRuleTypeOrder.indexOf()
}
```

**Review by dinujoh — CHANGES_REQUESTED** [review]

File: data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/util/SimpleExpressionMatcher.java (hunk 3 above)
Comment (sbayer55) [reply data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/util/SimpleExpressionMatcher.java:28]: Thanks for the idea @dinujoh. I'm not sure I completely understand your suggestion, in an earlier iteration of `getRuleIndex` I used `indexOf()`. Cases with subclasses did not work as expected.

A quick example:

```java
    private class NewContext extends DataPrepperExpressionParser.ExpressionContext {
        public NewContext(final ParserRuleContext parent, final int invokingState) {
            super(parent, invokingState);
        }
    }
    
    private int getIndex() {
        final List<Class<? extends ParseTree>> validRuleOrder = Arrays.asList(
                DataPrepperExpressionParser.ExpressionContext,
                DataPrepperExpressionParser.ConditionalExpressionContext,
                DataPrepperExpressionParser.RegexOperatorExpressionContext
        );
        
        validRuleOrder.indexOf(NewContext.class); // expected 0, will return -1
    }
```

Comment (dinujoh) [reply data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/util/SimpleExpressionMatcher.java:28]: For the subclass use case to work, you need to override the indexOf method:

```
List<Class<? extends ParseTree>> _validRuleOrder = new ArrayList<Class<? extends ParseTree>>(validRuleOrder) {
    @Override
    public int indexOf(Object item) {
        for (int x = 0; x < this.size(); x++) {
           if (this.get(x).isInstance(item)) {
              return x;
           }
        }
       return -1;
    }
};
```

Looks like validRuleOrder List data structure is used to find the index of the type/subtype of passed in item. We can define another data structure and abstract the operations. SimpleExpressionMatcher will be much cleaner and maintainable. 

```
public class RuleOrderTypeMatcher <T> {
  private final List<Class<? extends T>> ruleOrderByType;
  public RuleOrderTypeMatcher(final List<Class<? extends T>> validRuleOrder) { 
      List<Class<? extends T>> ruleOrderByType = new ArrayList<Class<? extends T>>(validRuleOrder) {
    @Override
    public int indexOf(Object item) {
        for (int x = 0; x < this.size(); x++) {
           if (this.get(x).isInstance(item)) {
              return x;
           }
        }
       return -1;
    }
   };
  }

 public boolean isValidRuleOrder(..) {
 ...
 }

 public boolean matchesLastNode(final T item) {
 ..
 }
}
```

Comment (dinujoh) [reply data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/util/SimpleExpressionMatcher.java:26]: Thanks for making this change.

**Review by dinujoh — APPROVED** [review]

**Review by sshivanii — APPROVED** [review]
Looks good, thanks for the updates.

**Pull request merged**